1-Draft/RFC00XX-Update-Error-View.md

```diff
@@ -0,0 +1,203 @@
+complete detailed view of the fully qualified error when desired.
+
+## Motivation
+
+The on-screen experience, when receiving an error message,
+is controlled through the views NormalView (the default) and CategoryView. These are user selectable
+through the preference variable $ErrorView.
+This RFC describes Changing '$ErrorView to an enumeration and adding one additional dynamic view
+to improve readability; 'ConciseView'
+
+- ConciseView - provides a concise error message suitable for new or occasional PowerShell users
+and a refactored view for advanced module builders. If the error is not from a script or parser
+error, then it's a single line error message. Otherwise, you get a multiline error message that
+contains the error and a pointer and error message showing where the error is in that line.
+If the terminal doesn't support Virtual Terminal, then vt100 color codes are not used.
+
+A comprehensive detailed view of the fully qualified error, including inner exceptions,
+will be provided by the `Resolve-ErrorRecord` cmdlet.
+
+'$ErrorView' shall contain the original views for backward compatibility and to
+lessen this breaking change. The view list is as follows:
+
+- ConciseView
+- NormalView
+- CategoryView
+
+## Specification
+
+The proposal is to add one new view to help improve error message comprehension
+based on user needs. For in-depth troubleshooting, a new cmdlet
+Resolve-ErrorRecord to provide detailed error information.
+
+__Key Design Considerations__
+
+1. To reduce confusion and improve debugging success for new and occasional users,
+error messages should call WriteErrorLine to produce a simplified message for interactive CLI users.
+
+- Simplified error message syntax from 'Message'. (See graphic below)
+
+```powershell
+PS C:\> Get-Childitem -Path c:\notreal
+Get-Childitem: Cannot find path ‘C:\notreal’ because it does not exist
+```
+
+2. To improve script debugging for advanced module builders and scripters,
+a refactored error view will be displayed. If the error is not from a script or parser
+error, then it's a single line error message. Otherwise, you get a multiline error message that
+contains the error and a pointer and error message showing where the error is in that line.
+
+- A new property ErrorAccentColor is added to support changing the accent color of the error message.
+If the terminal doesn't support Virtual Terminal, then vt100 color codes are not used.
+
+```powershell
+PS C:\> .\MyScript.ps1
+Get-ChildItem: in /Users/Username/GitHub/Errorview/script.test.ps1
+Line |
+15   | Get-ChildItem -Path c:\notreal
+     |                     ^^^ Cannot find path 'C:\notreal' because it does not exist.
+    * Help: Additional help information provided here
+```
+
+- An example is in the image below:
+
+![Message and Analytic](.\RFC00XX-Update-Error-View.png)
+
+3. A new cmdlet `Resolve-ErrorRecord` will produce comprehensive detailed
+view of the fully qualified error, including nested inner exceptions.
+
```

Comment:
Instead of defaulting to nesting inner exceptions, surface [GetBaseException](https://docs.microsoft.com/en-us/dotnet/api/system.exception.getbaseexception?view=netframework-4.8). This is the most inner exception and typically will be what people will want from my experience. I did this with SQL Server APIs and it gave me the best details without having to see all of the nested exceptions.

You can do this today by doing, `$_.Exception.GetBaseException()`.

A `[switch]` parameter can be used to resolve the full nested inner exceptions for those that still find value in showing all of them.

Reply:
With ConciseView done, you already have a view without base exception. This suggestion is to be additive by including the Base Exception in the next detailed view before going all in with nested exceptions.

Quite often, Base Exception is the same as the Outer Exception. The two exceptions could be compared to see if it should be surfaced or not for a "smarter" view. I don't know how that would impact performance though.

Reply:
At this time we are comfortable with providing the inner exceptions without the base exception, which seems redundant in many cases. We are also comfortable with the above work around in those cases if you need the base exception of $_.Exception.GetBaseException(). In the future, we can entertain additional views or can be sent as a PR.